docs/source/python_api/mlflow.data.rst

```diff
@@ -0,0 +1,158 @@
+mlflow.data
+============
+
+The ``mlflow.data`` module helps you record your model training and evaluation datasets to
+runs with MLflow Tracking, as well as retrieve dataset information from runs. It provides the
+following important interfaces:
+
+* :py:class:`Dataset <mlflow.data.dataset.Dataset>`: Represents a dataset used in model training or
+  evaluation, including features, targets, and metadata such as the dataset's name, digest (hash)
+  schema, profile, and source. You can log this metadata to a run in MLflow Tracking using
+  the :py:func:`mlflow.log_input()` API. ``mlflow.data`` provides APIs for constructing
+  :py:class:`Datasets <mlflow.data.dataset.Dataset>` from a variety of Python data objects, including
+  Pandas DataFrames (:py:func:`mlflow.data.from_pandas()`), NumPy arrays
+  (:py:func:`mlflow.data.from_numpy()`), Spark DataFrames (:py:func:`mlflow.data.from_spark()`
+  / :py:func:`mlflow.data.load_delta()`), and more.
+
+* :py:func:`DatasetSource <mlflow.data.dataset_source.DatasetSource>`: Represents the source of a
+  dataset. For example, this may be a directory of files stored in S3, a Delta Table, or a web URL.
+  Each :py:class:`Dataset <mlflow.data.dataset.Dataset>` references the source from which it was
+  derived. A :py:class:`Dataset <mlflow.data.dataset.Dataset>`'s features and targets may differ
+  from the source if transformations and filtering were applied. You can get the
+  :py:func:`DatasetSource <mlflow.data.dataset_source.DatasetSource>` of a dataset logged to a
+  run in MLflow Tracking using the :py:func:`mlflow.data.get_source()` API.
+
+The following example demonstrates how to use ``mlflow.data`` to log a training dataset to a run,
+retrieve information about the dataset from the run, and load the dataset's source.
+
+.. code-block:: python
+
+    import mlflow.data
+    import pandas as pd
+    from mlflow.data.pandas_dataset import PandasDataset
+
+    # Construct a Pandas DataFrame using iris flower data from a web URL
+    dataset_source_url = "http://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-red.csv"
+    df = pd.read_csv(dataset_source_url)
+    # Construct an MLflow PandasDataset from the Pandas DataFrame, and specify the web URL
+    # as the source
+    dataset: PandasDataset = mlflow.data.from_pandas(df, source=dataset_source_url)
```

Comment:
What is the purpose of annotating `dataset` with `PandasDataset`? As a user seeing this example, I wonder if it is necessary for the mlflow dataset feature to properly work.

Reply:
It's not necessary. The purpose is just to indicate that the return type of `from_pandas` is `PandasDataset`